activerecord/lib/active_record/associations/join_dependency.rb

```diff
@@ -248,22 +248,25 @@ def construct(ar_parent, parent, row, seen, model_cache, strict_loading_value)
               next
             end
 
-            key = aliases.column_alias(node, node.primary_key)
+            key = aliases.column_alias(node, node.primary_key || node.reflection.join_primary_key.to_s)
+
             id = row[key]
             if id.nil?
               nil_association = ar_parent.association(node.reflection.name)
               nil_association.loaded!
               next
             end
 
-            model = seen[ar_parent][node][id]
+            # Only cache when we have a primary key. This avoids issues when
+            # the non-unique foreign key is used as a fallback.
+            model = seen[ar_parent][node][id] if node.primary_key
 
             if model
```

Comment:
will `model` always be defined here if `node.primary_key` is nil?

I've pulled your code locally and ran the tests - and I believe the answer is yes. But it's difficult to verify from reading the code.

Reply:
Could you please explain a bit more? I tried a few things out but have seen failed tests in each case.

Reply:
`model` is a local variable belonging to the `construct` function, right?

If `model` isn't declared, `if model` will raise a NameError: `NameError (undefined local variable or method `model' for main:Object)`

Before changing:

```ruby
model = seen[ar_parent][node][id]
```

to:

```ruby
model = seen[ar_parent][node][id] if node.primary_key
```

we know for certain the `model` variable has been declared. But if `node.primary_key` is false/nil, then we're relying on `model` having been declared somewhere else.

Reply:
Ha! I never knew that.

> If you think this is too indirect, I'd be happy to change the code to be clearer.

While Ruby's behavior unexpected, I think the code's good as is.

Reply:
Great!